### ISSUE TITLE
REQUEST: Option for "Backspace as Undo"

### ISSUE THREAD

**Issue by mblais** [DrMerfy/vscode-overtype#20]
Please add a Settings option for "Backspace as Undo in Overtype mode" - so when in Overtype mode, hitting Backspace would simply undo the last character insertion, which would leave the line structure/spacing undisturbed.

**Comment by stkb** [comment]
Yeah I'd like to see this too. Yes you can just use ctrl-z or left arrow, but your brain is so wired to hitting the backspace key after making a typo that I do this all the time.

**Comment by blaumeise20** [comment]
I would rather like a new setting called "overtype.backspaceAction" setting:
- "delete": default behavior
- "undo": undo last character insertion
- "replace": separate setting "overtype.backspaceChar" where you can set the character to get inserted

**Comment by DominoPivot** [comment]
That would indeed be great. The "replace" option would actually replace the character currently in front of the caret, then move the caret back by 1 position. Example:

    12345  becomes 1 345
    12█45          1█345

Bikeshedding a little considering nobody seems to be working on this, but this might be worth considering:
- Should the extension use different terminology to avoid confusion between the "undo" action and the undo/redo commands?
- How would the undo/redo stack (Ctrl+Z) be affected by this operation?

**Comment by jkyeung** [comment]
There are three possible meanings for the backspace key in overtype mode that make sense to me personally:

1. Behave the same as backspace in insert mode (which is what it does now).
2. Behave the same as left arrow (which would be my preference).
3. Replace the character to the left of the cursor with a space, and then move the cursor left. (Or equivalently, do a backspace like in insert mode, and then insert a space but without moving the cursor.)

I really hate 1, because to me the entire point of overtype mode is that everything to the right of the cursor stays where it is. Though I hate the existing behavior, I think it makes sense to at least keep it as an option so that people who are used to it can continue to use it like that.

The difference between 2 and 3 is that 3 is "destructive", which at least conceptually makes some sense to me. It literally would be like typing a ***space***, except moving ***back***ward instead of forward.

While I think I understand the appeal of a single-character-undo behavior for the backspace key, I suspect it might be a little tricky to implement (because it's different than Ctrl+Z, which will undo an entire "batch" of keystrokes that VS Code has deemed to be one operation); and if you move the cursor using the arrow keys or mouse, and *then* press backspace, what should happen? The only sensible thing at that point would be one of the three choices I presented above.

***

**Addendum:** I think option 3 is better expressed as "behave like insert-mode backspace, except preserving the position of anything remaining to the right of the cursor". It's better not to refer specifically to deleting or inserting "a space", because the regular backspace can remove an entire level of indentation (consisting of a tab or multiple spaces), not just one space. So, "appropriate" _whitespace_ must be inserted to compensate for whatever was removed by backspace.

**Comment by MaddyGuthridge** [comment]
Building upon this, the one feature I had mentioned in #32 (which I closed since it is similar to this one) was having a similar option for the delete character (ie move the cursor forwards, since replacing with a space is covered by the spacebar)
